engine/table/src/main/java/io/deephaven/stream/TableStreamPublisherImpl.java

```diff
@@ -20,66 +20,99 @@
+import java.util.stream.Stream;
 
-class TablePublisherImpl implements StreamPublisher {
+class TableStreamPublisherImpl implements StreamPublisher {
 
     private final String name;
     private final TableDefinition definition;
+    private final Runnable onShutdownCallback;
     private final int chunkSize;
-    private StreamConsumer consumer;
+    private StreamToBlinkTableAdapter adapter;
 
-    TablePublisherImpl(String name, TableDefinition definition, int chunkSize) {
+    TableStreamPublisherImpl(String name, TableDefinition definition, Runnable onShutdownCallback, int chunkSize) {
         if (chunkSize <= 0) {
             throw new IllegalArgumentException("chunkSize must be positive");
         }
         this.name = Objects.requireNonNull(name);
         this.definition = Objects.requireNonNull(definition);
+        this.onShutdownCallback = Objects.requireNonNull(onShutdownCallback);
         this.chunkSize = chunkSize;
     }
 
     @Override
     public void register(@NotNull StreamConsumer consumer) {
-        if (this.consumer != null) {
+        if (this.adapter != null) {
             throw new IllegalStateException("Can not register multiple StreamConsumers.");
         }
-        this.consumer = Objects.requireNonNull(consumer);
+        if (!(consumer instanceof StreamToBlinkTableAdapter)) {
+            throw new IllegalStateException("TableStreamPublisher semantics only work with StreamToBlinkTableAdapter");
+        }
+        this.adapter = (StreamToBlinkTableAdapter) Objects.requireNonNull(consumer);
+    }
+
+    public TableDefinition definition() {
+        return definition;
+    }
+
+    public Table table() {
+        return adapter.table();
     }
 
     public void add(Table table) {
-        definition.checkMutualCompatibility(table.getDefinition());
+        // Note: we expect an exception in FillChunks / io.deephaven.engine.table.Table#getColumnSource if the table is
+        // incompatible instead of a heavier-weight
+        // io.deephaven.engine.table.TableDefinition checkCompatibility / checkMutualCompatibility
+        // we could do here.
         final FillChunks fillChunks = new FillChunks(table);
         try (final SafeCloseable ignored =
                 ExecutionContext.getContext().withUpdateGraph(table.getUpdateGraph()).open()) {
             ConstructSnapshot.callDataSnapshotFunction(
-                    TablePublisherImpl.class.getSimpleName() + "-" + name,
+                    TableStreamPublisherImpl.class.getSimpleName() + "-" + name,
                     ConstructSnapshot.makeSnapshotControl(false, table.isRefreshing(), (NotificationStepSource) table),
                     fillChunks);
         }
-        consumer.accept(fillChunks.outstandingChunks);
+        adapter.accept(fillChunks.outstandingChunks);
+    }
+
+    public void publishFailure(Throwable e) {
+        adapter.acceptFailure(e);
+    }
+
+    public boolean isAlive() {
+        return adapter.isAlive();
```

Comment:
I think this is silly to expose, and encourages the pattern of only supporting one consumer type. Better to have a boolean flagged in shutdown. Or leave that for the caller to do from shutdown.

Reply:
I somewhat disagree; at least, with respect to your desire to make onShutdownCallback nullable. If onShutdownCallback is null, the publisher needs another way to check if they are alive.

Even though you might think it's a crappy pattern, I think it will be a *common* pattern, so I'd still like to make it easy (instead of forcing the user to write their own boolean flagged shutdown callback).